lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+  if (D->isInvalid() || D->isImplicit())
+    return;
+
+  if (auto *VD = dyn_cast<ValueDecl>(D)) {
+    if (VD->getFormalAccess() != AccessLevel::Internal)
+      return;
+    if (!VD->isUsableFromInline())
+      return;
+  }
+
+  switch (D->getKind()) {
+  case DeclKind::Import:
+  case DeclKind::Extension:
+  case DeclKind::TopLevelCode:
+  case DeclKind::InfixOperator:
+  case DeclKind::PrefixOperator:
+  case DeclKind::PostfixOperator:
+  case DeclKind::PrecedenceGroup:
+  case DeclKind::Module:
+    llvm_unreachable("cannot appear in a type context");
+
+  case DeclKind::Param:
+  case DeclKind::GenericTypeParam:
+  case DeclKind::MissingMember:
+    llvm_unreachable("does not have access control");
+
+  case DeclKind::IfConfig:
+  case DeclKind::PoundDiagnostic:
+    // Does not have access control.
+  case DeclKind::EnumCase:
+    // Handled at the EnumElement level.
+  case DeclKind::Var:
+    // Handled at the PatternBindingDecl level.
+  case DeclKind::Destructor:
+    // Always correct.
+    return;
+
+  case DeclKind::PatternBinding: {
+    auto PBD = cast<PatternBindingDecl>(D);
+    bool isTypeContext = PBD->getDeclContext()->isTypeContext();
+
+    llvm::DenseSet<const VarDecl *> seenVars;
+    for (auto entry : PBD->getPatternList())
+    entry.getPattern()->forEachNode([&](const Pattern *P) {
+      if (auto *NP = dyn_cast<NamedPattern>(P)) {
+        // Only check individual variables if we didn't check an enclosing
+        // TypedPattern.
+        const VarDecl *theVar = NP->getDecl();
+        if (theVar->getFormalAccess() != AccessLevel::Internal)
+          return;
+        if (!theVar->isUsableFromInline())
+          return;
+        if (seenVars.count(theVar) || theVar->isInvalid())
+          return;
+
+        checkTypeAccess(TC, TypeLoc::withoutLoc(theVar->getType()),
+                        theVar, /*checkUsableFromInline=*/true,
+                        [&](AccessScope typeAccessScope,
+                            const TypeRepr *complainRepr,
+                            DowngradeToWarning downgradeToWarning) {
+          TC.diagnose(P->getLoc(),
+                      diag::pattern_type_not_usable_from_inline_inferred,
+                      theVar->isLet(),
+                      isTypeContext,
+                      theVar->getType());
+        });
+        return;
+      }
+
+      auto *TP = dyn_cast<TypedPattern>(P);
+      if (!TP)
+        return;
+
+      // FIXME: We need an access level to check against, so we pull one out of
+      // some random VarDecl in the pattern. They're all going to be the same,
+      // but still, ick.
+      const VarDecl *anyVar = nullptr;
+      TP->forEachVariable([&](VarDecl *V) {
+        seenVars.insert(V);
+        anyVar = V;
+      });
+      if (!anyVar)
+        return;
+      if (anyVar->getFormalAccess() != AccessLevel::Internal)
+        return;
+      if (!anyVar->isUsableFromInline())
+        return;
```

Comment:
This "internal and usable from inline" check shows up a lot. Maybe worth a helper function, even though it's so simple?

Reply:
Yeah. I'm going to drop the assertion from isUsableFromInline() that the access level is internal, and just return false if its not.